python/pyarrow/array.pxi

```diff
@@ -3573,17 +3595,20 @@ class FixedShapeTensorArray(ExtensionArray):
           ]
         ]
         """
-        if not obj.flags["C_CONTIGUOUS"]:
-            raise ValueError('The data in the numpy array need to be in a single, '
-                             'C-style contiguous segment.')
+
+        permutation = (-np.array(obj.strides)).argsort()
+        if not permutation[0] == 0:
+            raise ValueError('First stride needs to be largest to ensure that '
+                             'individual tensor data is contiguous in memory.')
 
         arrow_type = from_numpy_dtype(obj.dtype)
-        shape = obj.shape[1:]
-        size = obj.size / obj.shape[0]
+        shape = np.take(obj.shape, permutation)
+        values = np.lib.stride_tricks.as_strided(
+            obj, shape=(obj.size,), strides=(obj.dtype.itemsize,))
```

Comment:
Is this different from the ravel that was used before? (or maybe you need `ravel(.., order="K")` to just flatten following the memory layout

Reply:
The reason for the change was that ravel states in the doc: `[..] A copy is made only if needed. [..]` while `as_strided` only creates a view. I would prefer `ravel` as it's less error prone but I can't immediately see from [source](https://github.com/numpy/numpy/blob/v1.26.0/numpy/core/fromnumeric.py#L1768-L1874) when it would incur a copy.

Reply:
In general it will make a copy when the memory layout changes, but by specifying `order="K"`, you say to use the original memory layout, which I think means it will always be zero-copy

Reply:
I've not seen zero-copy specifically promised for `order="K"`, but it seem it's meant for that. Changed to `ravel`.